src/main/java/org/radarbase/appserver/service/storage/S3StorageService.java

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ *  *  Copyright 2024 The Hyve
+ *  *
+ *  *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  *  you may not use this file except in compliance with the License.
+ *  *  You may obtain a copy of the License at
+ *  *
+ *  *    http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  *  Unless required by applicable law or agreed to in writing, software
+ *  *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  *  See the License for the specific language governing permissions and
+ *  *  limitations under the License.
+ *
+ */
+
+package org.radarbase.appserver.service.storage;
+
+import io.minio.PutObjectArgs;
+import lombok.extern.slf4j.Slf4j;
+import org.radarbase.appserver.config.S3StorageProperties;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+import org.springframework.web.multipart.MultipartFile;
+
+import javax.annotation.PostConstruct;
+
+
+@Slf4j
+@Service
+@ConditionalOnExpression("${file-upload.enabled:false} and 's3' == '${storage.type:}'")
+public class S3StorageService extends RandomUuidFilenameStorageService implements StorageService {
+
+    @Autowired
+    private S3StorageProperties s3StorageProperties;
+
+    @Autowired
+    private MinioClientInitializer bucketClient;
+    private String subPath = "";
+
+    @PostConstruct
+    public void init() {
+        if (s3StorageProperties.getSubPath() != null) {
+            subPath = s3StorageProperties.getSubPath().replaceAll("^/|/$", "");
+            if (!subPath.isEmpty()) {
+                subPath = subPath + "/";
+            }
+        }
+    }
+
+    public String store(MultipartFile file, String projectId, String subjectId, String topicId) {
+        Assert.notNull(file, "File must not be null");
+        Assert.notEmpty(new String[]{projectId, subjectId, topicId}, "Project, subject and topic IDs must not be empty");
```

Comment:
It is probably better to use specific [exceptions](https://github.com/RADAR-base/RADAR-Appserver/tree/master/src/main/java/org/radarbase/appserver/exception) with HTTP codes and handle them with [Exception handler](https://github.com/RADAR-base/RADAR-Appserver/blob/dev/src/main/java/org/radarbase/appserver/exception/handler/ResponseEntityExceptionHandler.java).

Reply:
I do not quite understand what you mean here. In the Service layer it does ont make sense to me to throw HTTP-level exceptions. The Assert in the service throws an IllegalArgumentException which is fine I think. Or ar e you not proposing to send HTTP-level exceptions from the service?

Reply:
These will be relayed to the app as an HTTP response, so it makes sense to have a specific error message in case they are useful. You can create new [exceptions](https://github.com/RADAR-base/RADAR-Appserver/tree/master/src/main/java/org/radarbase/appserver/exception) like `InvalidFileDetailsException extends IllegalArgumentException` and annotate with `@ResponseStatus(HttpStatus.XYZ)`. This can then be converted to a proper response for the app in [Exception handler](https://github.com/RADAR-base/RADAR-Appserver/blob/dev/src/main/java/org/radarbase/appserver/exception/handler/ResponseEntityExceptionHandler.java). Alternative is to catch the assertion error in the controller and raise these exceptions there, but I prefer doing it in the service (or where the exception is raised) as you can include more context. Essentially, you are not sending HTTP responses from the service but raising usual exceptions and configuring spring to convert to responses (this avoids handling them manually and keeps all the exception handling code neatly in one place).

Reply:
Ah ok, I think I understand. Could you please review my last commit to see whether this is what you intended?